torchtune/_cli/download.py

```diff
@@ -159,3 +202,91 @@ def _download_cmd(self, args: argparse.Namespace) -> None:
             *list(Path(true_output_dir).iterdir()),
             sep="\n",
         )
+
+    def _download_from_kaggle(self, args: argparse.Namespace) -> None:
+        """Downloads a model from the Kaggle Model Hub."""
+
+        # Note: Kaggle doesn't actually use the "repository" terminology, but we still reuse args.repo_id here for simplicity
+        model_handle = args.repo_id
+        self._validate_kaggle_model_handle(model_handle)
+
+        # Default output_dir is `/tmp/<model_name>`, stripping the publisher name to align with HF output_dir format
+        output_dir = args.output_dir
+        if output_dir is None:
+            model_name = "/".join(args.repo_id.split("/")[1:]).lower()
+            output_dir = Path("/tmp") / Path(model_name)
+
+        if bool(args.kaggle_username) ^ bool(args.kaggle_api_key):
+            self._parser.error(
+                "Both your Kaggle username and api key are needed to authenticate to Kaggle, but only one was provided."
+            )
+        elif bool(args.kaggle_username) and bool(args.kaggle_api_key):
+            set_kaggle_credentials(args.kaggle_username, args.kaggle_api_key)
+
+        try:
+            kaggle_output_path = model_download(model_handle)
+        except KaggleApiHTTPError as e:
+            if e.response.status_code in {
+                HTTPStatus.UNAUTHORIZED,
+                HTTPStatus.FORBIDDEN,
+            }:
+                self._parser.error(
+                    "It looks like you are trying to access a gated model. Please ensure you "
+                    "have access to the model and have provided the proper Kaggle credentials "
+                    "using the options `--kaggle-username` and `--kaggle-api-key`. You can also "
+                    "set these to environment variables as detailed in "
+                    "https://github.com/Kaggle/kagglehub/blob/main/README.md#authenticate."
+                )
+            elif e.response.status_code == HTTPStatus.NOT_FOUND:
+                self._parser.error(
+                    f"'{model_handle}' not found on the Kaggle Model Hub."
+                )
+            tb = traceback.format_exc()
+            msg = f"Failed to download {model_handle} with error: '{e}' and traceback: {tb}"
+            self._parser.error(msg)
+        except Exception as e:
+            tb = traceback.format_exc()
+            msg = f"Failed to download {model_handle} with error: '{e}' and traceback: {tb}"
+            self._parser.error(msg)
+
+        try:
+            # manually copy contents as workaround until kagglehub supports explicit output_dir specification
+            true_output_dir = copytree(
```

Comment:
This work can be done by our team, or quick PR on our side. It will be easier to release changes on our side than to merge into the torch tune lib. Updates on their side will only have to be update in version number.

Reply:
quick update: I removed the `copytree` shenanigans entirely. After talking with @rosbo, I think the `output-dir` behavior is more involved than we previously expected. For now, I changed torchtune to just `warn` the user when they provided `--output-dir` when downloading from kaggle